src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -312,19 +335,84 @@ def _prepare_inits(self) -> Set[str]:
         ns_package.write(b'__import__("pkg_resources").declare_namespace(__name__)  # type: ignore[attr-defined]')
         ns_package.flush()
         for missing_init_file in missing_init_files:
+          self._all_added_sources_resources.append(Path(missing_init_file))
           self._builder.add_source(filename=ns_package.name, env_filename=missing_init_file)
     return missing_init_files
 
   def set_emit_warnings(self, emit_warnings):
     self._builder.info.emit_warnings = emit_warnings
 
+  def _shuffle_original_build_info_into_ptex(self):
+    """Create a "dehydrated" ptex file without any of its requirements, and specify that in two *-INFO files.
+
+    See pex_bootstrap/_ptex_launcher.py for details of how these files are used.
+    """
+    ipex_info = self._builder.info.copy()
+    orig_chroot = self._builder.chroot()
+    # Replace the original builder with a new one, and just pull files from the old chroot.
+    self._builder = PEXBuilder(interpreter=self._builder.interpreter)
+
+    # PTEX-INFO: Contains resolver settings and source files to copy over into the dehydrated .ipex
+    #            file.
+    all_code = [str(src) for src in self._all_added_sources_resources]
+    for src in all_code:
+      # NB: Need to add under 'source' label for `self._prepare_inits()` to pick it up!
+      self._builder.chroot().copy(os.path.join(orig_chroot.path(), src), src, label='source')
+
+    python_repos = self._python_repos_subsystem
+
+    resolver_settings = dict(
+      indexes=list(python_repos.indexes),
+      find_links=list(self._all_find_links),
+    )
+
+    ptex_info = dict(
+      code=all_code,
+      resolver_settings=resolver_settings,
+    )
+    with temporary_file(permissions=0o644) as ptex_info_file:
+      ptex_info_file.write(json.dumps(ptex_info).encode())
+      ptex_info_file.flush()
+      self._builder.add_resource(filename=ptex_info_file.name, env_filename='PTEX-INFO')
+
+    # IPEX-INFO: The original PEX-INFO, which should be the PEX-INFO in the hydrated .pex file that
+    #            is generated when the .ipex is first executed.
+    with temporary_file(permissions=0o644) as ipex_info_file:
+      ipex_info_file.write(ipex_info.dump().encode())
+      ipex_info_file.flush()
+      self._builder.add_resource(filename=ipex_info_file.name, env_filename='IPEX-INFO')
+
+    # _ptex_launcher.py: The special bootstrap script to hydrate the .ipex with the fully resolved
+    #                    requirements when it is first executed.
+    parent_module = 'pants.backend.python.subsystems'
```

Comment:
```suggestion
    parent_module = os.path.dirname(__name__)
```
maybe?

Reply:
I added `module_dirname()` to `pants.util.strutil` to do this!